Enable passing arguments on queue creation + introduce buildTable

iainb opened the pull request:
queue.bind was not passing options.arguments during queue creation
meaning all queues were created without any arguments.

Introduced buildTable method to allow supporting multiple data types
when creating queues and passing attributes when publishing messages.


Note: I realise this needs tests, it works when I've been running it locally. Would be nice to get feedback before investing that time.

Edit: package.json needs updating too with a version bump

glenjamin commented:
Hi! Thanks for the PR.

I like buildTable - it's much cleaner than the previous approach. If you make sure it can handle an empty object, you can remove some of the if checks before calling.

As for passing arguments to bind - I think this needs to be a differently named option, as the arguments to declare & bind should be different?

iainb commented:
Hi Glen,

I've updated buildTable so it can handle 'undefined' as input (it returns an empty object, which bramqp can handle ok) and removed the checks before it is called.

With regards to the arguments to queue.bind, you're right the 'arguments' attribute for queue.declare and queue.bind are different. I've updated the code and documentation to reflect this.

Note: I squashed my changes and re-wrote the commit message, this is a fairly small pull req so hopefully it won't be hard to see what has changed.

glenjamin commented:
Great stuff! A follow-up PR with tests would also be much appreciated :)
